hashpage.go

```diff
@@ -0,0 +1,252 @@
+func (es *ExploreServer) parseTransactions(hashes []crypto.Hash) ([]byte, error) {
+	var page []byte
+
+	// Request all the other things to be viewed
+	for _, hash := range hashes {
+		parsed, err := es.parseTransaction(hash)
+		if err != nil {
+			return nil, err
+		}
+		page = append(page, parsed...)
+	}
+	return page, nil
+}
+
+// blockPage formats information about a block for viewing
+func (es *ExploreServer) blockPage(w http.ResponseWriter, blockJSON []byte) {
+	var b modules.BlockResponse
+	err := json.Unmarshal(blockJSON, &b)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	blockSummaries, err := es.apiGetBlockData(b.Height, b.Height+1)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// blockRoot is defined in template.go
+	page, err := es.parseTemplate("block.html", blockRoot{
+		Block:  b.Block,
+		Height: b.Height,
+		Target: blockSummaries[0].Target,
+		Size:   blockSummaries[0].Size,
+	})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	w.Write(page)
+}
+
+// txPage formats information about a transaction in a format suitable
+// for human viewing
+func (es *ExploreServer) txPage(w http.ResponseWriter, txJSON []byte) {
+	var t modules.TransactionResponse
+	err := json.Unmarshal(txJSON, &t)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Parse the main transaction template
+	page, err := es.parseTemplate("transaction.html", t)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	w.Write(page)
+}
+
+// outputPage formats an outputTransactions struct for use in viewing
+func (es *ExploreServer) outputPage(w http.ResponseWriter, outJSON []byte, oID []byte) {
+	var ot modules.OutputResponse
+	err := json.Unmarshal(outJSON, &ot)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	var outputID crypto.Hash
+	copy(outputID[:], oID[:])
+
+	// Parse the main output template
+	page, err := es.parseTemplate("output.html", outputRoot{
+		OutputTx: ot.OutputTx,
+		InputTx:  ot.InputTx,
+		OutputID: outputID,
+	})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	w.Write(page)
+}
+
+// addressPage formats the list of transactions that an address
+// participated in for human consumption
+func (es *ExploreServer) addressPage(w http.ResponseWriter, addrJSON []byte, address []byte) {
+	var ad modules.AddrResponse
+	err := json.Unmarshal(addrJSON, &ad)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Parse header template
+	page, err := es.parseTemplate("address.html", addressRoot{
+		Txns: ad.Txns,
+		Addr: address,
+	})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	w.Write(page)
+}
+
+func (es *ExploreServer) contractPage(w http.ResponseWriter, fcJSON []byte, fcid []byte) {
+	var fi modules.FcResponse
+	err := json.Unmarshal(fcJSON, &fi)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	var FcID types.FileContractID
+	copy(FcID[:], fcid[:])
```

Comment:
types.FileContractID is a a crypto.Hash which is a byte array. This conversion requires copying iirc.